tests/schema/tests.py

```diff
@@ -2391,6 +2395,338 @@ def test_text_field_with_db_index(self):
         assertion = self.assertIn if connection.features.supports_index_on_text_field else self.assertNotIn
         assertion('text_field', self.get_indexes(AuthorTextFieldWithIndex._meta.db_table))
 
+    @skipUnlessDBFeature('supports_expression_indexes')
+    def test_func_index(self):
+        """
+        Tests creating a functional index
+        """
+
+        # Create the table
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+            editor.create_model(Book)
+
+        # Define the index
+        index_name = 'lower_func_idx'
+        index = Index(fields=[Lower('title').desc()], name=index_name)
+
+        # Add the index
+        with connection.schema_editor() as editor:
+            editor.add_index(Book, index)
+            sql = index.create_sql(Book, editor)
+        self.assertIn(index_name, self.get_constraints(Book._meta.db_table))
+
+        sql = str(sql).upper()
+        self.assertIn('LOWER(', sql)
+        self.assertIn('TITLE', sql)
+        self.assertIn(index_name.upper(), sql)
+        self.assertIn('DESC', sql)
+
+        # Drop the index
+        with connection.schema_editor() as editor:
+            sql = index.remove_sql(Book, editor)
+            editor.remove_index(Book, index)
+        sql = str(sql).upper()
+        self.assertIn(index_name.upper(), sql)
+        self.assertNotIn(index_name, self.get_constraints(Book._meta.db_table))
+
+    @skipUnlessDBFeature('supports_expression_indexes')
+    def test_composite_func_index(self):
+        """
+        Test creating a composite functional index
+        """
+        # Create the table
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+
+        # Define the index
+        index_name = 'lower_upper_func_idx'
+        index = Index(fields=[Lower('name'), Upper('name')], name=index_name)
+
+        # Add the index
+        with connection.schema_editor() as editor:
+            editor.add_index(Author, index)
+            sql = index.create_sql(Author, editor)
+
+        # Ensure the index is there
+        self.assertIn(index_name, self.get_constraints(Author._meta.db_table))
+
+        # Check that the vital SQL parts are there
+        sql = str(sql).upper()
+        self.assertIn('LOWER(', sql)
```

Comment:
Could `Columns` or some other DDL class be used? I think this was originally written before the DDL classes were introduced.

Reply:
The new `Expressions` DDL class made this a lot easier 👍